src/expr/src/relation/mod.rs

```diff
@@ -1739,11 +1740,56 @@ impl MirRelationExpr {
         false
     }
 
+    /// Given the ids and values of a LetRec, it computes the subset of ids that are used across
+    /// iterations. These are those ids that have a reference before they are defined, when reading
+    /// all the bindings in order.
+    ///
+    /// For example:
+    /// ```SQL
+    /// WITH MUTUALLY RECURSIVE
+    ///     x(...) AS f(z),
+    ///     y(...) AS g(x),
+    ///     z(...) AS h(y)
+    /// ...;
+    /// ```
+    /// Here, only `z` is returned, because `x` and `y` are referenced only within the same
+    /// iteration.
+    ///
+    /// Note that if a binding references itself, that is also returned.
+    pub fn recursive_ids(
+        ids: &Vec<LocalId>,
+        values: &Vec<MirRelationExpr>,
+    ) -> Result<BTreeSet<LocalId>, RecursionLimitError> {
+        let mut used_across_iterations = BTreeSet::new();
+        let mut defined = BTreeSet::new();
+        for (binding_id, value) in itertools::zip_eq(ids.iter(), values.iter()) {
+            value.visit_post(&mut |expr| {
+                if let MirRelationExpr::Get {
+                    id: Local(get_id), ..
+                } = expr
+                {
+                    // If we haven't seen a definition for it yet, then this will refer
+                    // to the previous iteration.
+                    // The `ids.contains` part of the condition is needed to exclude
+                    // those ids that are not really in this LetRec, but either an inner
+                    // or outer one.
+                    if !defined.contains(get_id) && ids.contains(get_id) {
+                        used_across_iterations.insert(*get_id);
+                    }
+                }
+            })?;
+            defined.insert(*binding_id);
+        }
+        Ok(used_across_iterations)
+    }
+
     /// Replaces `LetRec` nodes with a stack of `Let` nodes.
     ///
     /// In each `Let` binding, uses of `Get` in `value` that are not at strictly greater
     /// identifiers are rewritten to be the constant collection.
     /// This makes the computation perform exactly "one" iteration.
+    ///
+    /// This was used only temporarily while developing `LetRec`.
```

Comment:
What does this comment imply? Can we remove it once all transforms are recursion safe?

Reply:
It's currently not used, so maybe we should already remove it. It was being used when the rendering was not fully implemented yet.